src/ssl/helper.cc

```diff
@@ -126,22 +119,26 @@ void Ssl::Helper::Submit(CrtdMessage const & message, HLPCB * callback, void * d
     SBuf rawMessage(message.compose().c_str()); // XXX: helpers cannot use SBuf
     rawMessage.append("\n", 1);
 
-    const auto pending = TheGeneratorRequests.find(rawMessage);
-    if (pending != TheGeneratorRequests.end()) {
-        pending->second->emplace(callback, data);
-        debugs(83, 5, "collapsed request from " << data << " onto " << *pending->second);
-        return;
+    if (ssl_crtd) {
+        const auto pending = ssl_crtd->generatorRequests.find(rawMessage);
+        if (pending != ssl_crtd->generatorRequests.end()) {
+            pending->second->emplace(callback, data);
+            debugs(83, 5, "collapsed request from " << data << " onto " << *pending->second);
+            return;
+        }
     }
 
     GeneratorRequest *request = new GeneratorRequest;
     request->query = rawMessage;
     request->emplace(callback, data);
-    TheGeneratorRequests.emplace(request->query, request);
     debugs(83, 5, "request from " << data << " as " << *request);
     // ssl_crtd becomes nil if Squid is reconfigured without SslBump or
     // certificate generation disabled in the new configuration
-    if (ssl_crtd && ssl_crtd->trySubmit(request->query.c_str(), HandleGeneratorReply, request))
```

Comment:
AFAICT, something went wrong during refactoring because added generator requests are now never removed from ssl_crtd->generatorRequests.

I understand that it used to be easy to guarantee the request presence in the static TheGeneratorRequests collection, and that we can no longer do that. I believe the best solution here is to throw an exception when ssl_crtd is missing instead of trying to handle its disappearance quietly, with a misleading "Temporary network problem" error.

The "ssl_crtd becomes nil if Squid is reconfigured without SslBump or certificate generation disabled in the new configuration" comment above misleads: Our caller _must_ (and does) check that dynamic generation of host certificates is enabled (i.e. port->secure.generateHostCertificates is true) -- the caller has no right to use the helper without that configuration option. When it is enabled and USE_SSL_CRTD is true, the admin wants this method to generate the host certificate . If the method cannot do that, it is an error!

How can ssl_crtd be nil inside Ssl::Helper::Submit() (i.e. with true USE_SSL_CRTD and secure.generateHostCertificates)? I am not sure. Please investigate whether that is possible during some late reconfiguration stages and/or after helper-killing helper errors. The outcome of that investigation may affect how we handle the error described above. For example, we may indeed throw (as sketched above) or do something else to avoid catch ERROR in the caller (e.g., check ssl_crtd in the caller or return false from this method).

Reply:
> whether that is possible

I don't think it is possible:

* On reconfiguration, Ssl::Helper::Reconfigure() calls Ssl::Helper::Shutdown(), that sets ssl_crtd to nil and then immediately  Ssl::Helper::Init(), that assigns ssl_crtd to a new helper. In other words, there is no situation when ssl_crtd remains nil until some context switch (allowing Ssl::Helper::Submit() be called).
* On helper crash, ssl_crtd remains unchanged, its shared copy passed to a new Session object when Session::HelperServerClosed() calls helperOpenServers().

Reply:
> On reconfiguration, Ssl::Helper::Reconfigure() calls Ssl::Helper::Shutdown(), that sets ssl_crtd to nil and then immediately Ssl::Helper::Init(), that assigns ssl_crtd to a new helper.

However, Ssl::Helper::Init() may exit without creating/assigning a new helper, leaving ssl_crtd nil. Our caller -- ConnStateData::getSslContextStart() -- checks that dynamic generation of host certificates is enabled, but it does not check flags.tunnelSslBumping. Moreover, I suspect that getSslContextStart() may be checking _stale_ port configuration in some cases (the one before reconfiguration)[^1]. Thus, AFAICT, we may be called after reconfiguration made ssl_crtd nil.

[^1]: If that is true, then simply adding a flags.tunnelSslBumping check to the caller would not really help. Bugs notwithstanding, the true state of flags.tunnelSslBumping (in the port configuration store/used by the transaction) is essentially implied by that transaction getting to the getSslContextStart() step.

That "ssl_crtd becomes nil if Squid is reconfigured without SslBump or certificate generation disabled in the new configuration" comment is arguably misleading or imprecise, but, AFAICT, it is not flat out wrong: We _can_ get here with a nil ssl_crtd.

When an pre-reconfiguration transaction needs ssl_crtd that has been removed by reconfiguration, is that a Squid bug or an admin-created problem? The answer will determine what/how we throw. In either case, the existing caller will log an ERROR and "fall through to do blocking in-process generation".

I lean towards a "Squid bug" conclusion: A transaction that knows it may need ssl_crtd in the foreseeable future should keep a pointer to that helper (like it keeps a pointer to the port configuration), preserving the helper across reconfigurations. In a sense, that helper is "attached to" the port configuration(s) that requires it. I doubt we should fix this bug in this PR (unless the fix is really trivial). This bug can and probably should be fixed in another/dedicated PR. For now, I suggest that we just throw along these lines:

```C++
        // A before-reconfiguration transaction may be using a PortCfg that
        // requires certificate generation when reconfiguration makes ssl_crtd
        // nil. TODO: Preserve the helper for such transactions.
        throw TextException("Squid BUG: A transaction-required ssl_crtd helper disappeared during reconfiguration", Here());
```

> On helper [process] crash, ssl_crtd remains unchanged

I agree. Only Ssl::Helper::Shutdown() makes ssl_crtd nil, and that function is only called during reconfiguration and upon SquidShutdown() AFAICT.

Reply:
>  comment is arguably misleading

If Ssl::Helper::Submit() may be called without SslBump configured, then 'reconfiguration' should not matter, i.e., we can get there having a port with generate-host-certificates but without ssl-bump.

Reply:
> we can get there having a port with generate-host-certificates but without ssl-bump.

That is probably true, but is also probably irrelevant in the current code: If a port configuration does not have ssl-bump configured, then the transaction using that port configuration is not going to get to Ssl::Helper::Submit(), regardless of the generate-host-certificates settings.

I am not sure why you made that comment though, so my response may not address your actual concern.

Reply:
I am just trying to understand the difference between configuration and reconfiguration in this context. If, as you say, the transaction on a port without ssl-bump does not call Ssl::Helper::Submit(), why this should happen after reconfiguration?

Reply:
> If, as you say, the transaction on a port without ssl-bump does not call Ssl::Helper::Submit(), why this should happen after reconfiguration?

```C++
void
ConnStateData::getSslContextStart()
{
    if (port->secure.generateHostCertificates) { // after reconfiguration, this is not necessarily the _current_ port configuration
        ...
        Ssl::Helper::Submit(request_message, sslCrtdHandleReplyWrapper, this);
```

```C++
class Server
{
    ...
    /// Squid listening port details where this connection arrived.
    AnyP::PortCfgPointer port;
```

Unlike configuration that happens before any transaction may start, reconfiguration creates a situation where a pre-reconfiguration transaction still has/uses old port configuration and the corresponding transaction state (that, together, used to imply, for example, the existence of ssl_crtd) while Ssl::Helper code uses new/current configuration (where ssl_crtd is, for example, gone because ssl-bump was disabled). Does this clarify?

Reply:
> something went wrong during refactoring 

Fixed at 05cbcce (after merging recent master changes).

>  suggest that we just throw 

Done (98d2d9a).